Intersect.Client.Core/Entities/Resource.cs

```diff
@@ -175,15 +229,48 @@ private void OnAnimationDisposedOrFinished(IAnimation animation)
         animation.Finished -= OnAnimationDisposedOrFinished;
     }
 
-    private void UpdateFromDescriptor(ResourceDescriptor? descriptor)
+    public void UpdateCurrentState()
     {
-        if (descriptor == null)
+        if (Descriptor == default)
         {
             return;
         }
 
-        var updatedSprite = IsDead ? descriptor.Exhausted.Graphic : descriptor.Initial.Graphic;
-        _sprite = updatedSprite;
+        var graphicStates = Descriptor.States;
+        var currentHealthPercentage = Math.Floor((float)Vital[(int)Enums.Vital.Health] / _maximumHealthForStates * 100);
+
+        if (_currentState is { } currentState &&
+            currentHealthPercentage >= _currentState?.MinimumHealth &&
+            currentHealthPercentage <= _currentState?.MaximumHealth
+        )
+        {
+            return;
+        }
+
+        currentState = graphicStates.Values.FirstOrDefault(
+            s => currentHealthPercentage >= s.MinimumHealth && currentHealthPercentage <= s.MaximumHealth
+        );
+
+        // dispose animation if animation is not used in this current state,
+        // but the previous state was an animation
+        if (
+            _currentState?.TextureType == ResourceTextureSource.Animation &&
+            _stateAnimation != default &&
+            currentState?.TextureType != ResourceTextureSource.Animation
```

Comment:
This does not need to check `_stateAnimation`, just use `?` when calling `Dispose()`